### PR TITLE
Queue interop producer\provider

### PR DESCRIPTION
**Pull request by makasim**
The PR adds provider\publisher that can work with any [queue interop](https://github.com/queue-interop/queue-interop) compatible transport.

TODO:

* [x] add tests
* [x] add doc

I welcome everyone who is interested in MQs to join the queue interop group

### CODE REVIEW

**Comment by olaurendeau** [comment]
Hello @makasim, queue-interop look like a great idea !
capture d ecran 2017-07-19 a 15 11 00
Looking at compatible queue systems list (and not being an expert at all), I'm just wondering : 
AMQP use the notion of routing_key, notion that you didn't implement in your producer because obviously it's not an avery queue system feature. Some Processors will not be compatible with queue-interop for example the retry one https://github.com/swarrot/swarrot/tree/master/src/Swarrot/Processor/Retry

As swarrot is theoretically built to be queue agnostic, but currently only support AMQP, do we need a mechanism explaining / constraining those limitations if swarrot have ambition to support other queueing systems ?

**Comment by makasim** [comment]
The AMQP implementation  [supports the ability to publish messages on the exchange with routing key set](https://github.com/php-enqueue/enqueue-dev/blob/master/pkg/amqp-ext/AmqpProducer.php#L57). It has to be set on the topic object with `setRoutingKey` method. But the method is not a part of the speck (neither queue-interop nor origin JMS). 

If the `$key` is used to publish the message to a queue with the same then it could be implemented. The amqp producer can publish messages on a queue without the need to have explicit exchange. 

```php
<?php
    public function publish(Message $message, $key = null)
    {
        // ....

        $this->producer->send(
            $key ? $this->context->createQueue($key) : $this->topic,
            $this->context->createMessage($message->getBody(), $properties, $headers)
        );
    }
```

**Comment by olaurendeau** [comment]
> The amqp producer can publish messages on a queue without the need to have explicit exchange.
Indeed and in this case routing_key will be used to define the destination queue name. 

```
$this->producer->send(
            $key ? $this->context->createQueue($key) : $this->topic,
            $this->context->createMessage($message->getBody(), $properties, $headers)
        );
```
If I'm not mistaken this would use the routing_key to address the queue having the name of the routing_key value ? So this is not supporting AMQP bindings.

For RPC it makes perfect sense as discussed here https://github.com/swarrot/SwarrotBundle/pull/94

But in the example of the Retry processor a message is publish on an exchange and the routing_key is used to dispatch the message on the proper queue through the AMQP bindings. So this processor would not be compatible with Queue interop, which would maybe be fine if we had a mechanism constraining processors to a set of Publishers / Providers, meanwhile I assume that specifying it in documentation would be enough 😃

**Comment by olaurendeau** [comment]
Yes this is precisely what I wanted to emphasis by the example of https://github.com/swarrot/SwarrotBundle/pull/94 :)

For the RetryProcessor the `amp.direct` + queue name as routing key feature will not do the trick.
Routing keys will be a pattern like `retry_%attempt` and the resulting key `retry_10`, `retry_9`, ... , `retry_1` that's why a binding `retry_#` between the retry exchange and the destination queue is necessary.

Anyway my point is simply that the Queue Interop Publisher / Provider documentation should precise that RetryProcessor will not be compatible :)

**Comment by makasim** [comment]
@olaurendeau finally got it. I'll mention it in the docs

**Comment by makasim** [comment]
btw we are thinking of building an amqp-interop on top of queue-interop. That would make it possible to support such AMQP specific cases without coupling to implementation.

**Comment by olaurendeau** [comment]
Make sense :) And there is others protocols providing specific features, like gearman.
Do you plan to move php-enqueue/* repos to queue-interop ?

**Comment by makasim** [comment]
> Do you plan to move php-enqueue/* repos to queue-interop ?

I dont think it worth it.

**Comment by makasim** [comment]
I am finished. Could you please have a look?

File: src/Swarrot/Processor/InteropProxyProcessor.php [hunk 1]
```diff
@@ -0,0 +1,47 @@
...
+
+    /**
+     * {@inheritdoc}
+     */
+    public function process(PsrMessage $message, PsrContext $context)
+    {
+        $properties = $message->getHeaders();
+        $properties['headers'] = $message->getProperties();
+
+        $this->swarrotProcessor->process(new Message($message->getBody(), $properties), $this->options);
```
Comment (olaurendeau) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: Why is this proxy useful as `InteropMessageProvider::get` will return a `Swarrot\Broker\Message` ?

File: tests/Swarrot/Processor/InteropProxyProcessorTest.php [hunk 2]
```diff
@@ -0,0 +1,207 @@
...
+    public function testInstance()
+    {
+        $swarrotProcessor = $this->prophesize('Swarrot\Processor\ProcessorInterface');
+        $this->assertInstanceOf(
+            'Interop\Queue\PsrProcessor',
+            new InteropProxyProcessor($swarrotProcessor->reveal())
+        );
+    }
+
+    public function test_get_with_messages_in_queue_return_message()
```
Comment (olaurendeau) [inline tests/Swarrot/Processor/InteropProxyProcessorTest.php]: Not sure about test name :)

File: tests/Swarrot/Processor/InteropProxyProcessorTest.php [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+
+        $processor = new InteropProxyProcessor($swarrotProcessor->reveal(), ['fooOpt' => 'fooOptVal']);
+
+        $result = $processor->process($message->reveal(), $context->reveal());
+
+        $this->assertSame(PsrProcessor::ACK, $result);
+
+
+    }
+//
```
Comment (olaurendeau) [inline tests/Swarrot/Processor/InteropProxyProcessorTest.php]: Maybe remove those commented lines ?

File: README.md [hunk 4]
```diff
@@ -105,6 +105,49 @@ Here is an illustration to show you what happens when this order is used:
...
+use Swarrot\Broker\Message;
+
+/** @var \Interop\Queue\PsrContext $context */
+
+$messagePublisher = new InteropMessagePublisher($context, 'aTopic');
+
+$messagePublisher->publish(new Message('aBody'));
+```
+
+_**Note: ** The second argument `key` is not supported at the moment._
```
Comment (olaurendeau) [inline README.md]: Maybe the place to also inform people that RetryProcessor is not supported by queue-interop publisher ?

File: src/Swarrot/Processor/InteropProxyProcessor.php (hunk 1 above)
Comment (makasim) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: @olaurendeau sorry, I dont get it.

**Comment by odolbeau** [comment]
Hello @makasim.
Thanks for this PR. I won't be able to review it correctly before 2-3 weeks as I don't have my computer with me at the moment.
Anyway it looks good for me and it's definitively a good thing to be compatible with queue-interop!
@olaurendeau feel free to merge my PR for PHP 7.1. You should be able to tag a new version. :) BTW, don't wait for me to merge this PR as soon as you think it's ready.

I'll take some time to dig into queue-interop group / repositories ASAP.

File: src/Swarrot/Processor/InteropProxyProcessor.php (hunk 1 above)
Comment (olaurendeau) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: I don't see the point of the InteropProxyProcessor :)

**Comment by olaurendeau** [comment]
@makasim Instead of changing .travis.yml could you simply rebase from master ? :)
Thanks !

File: src/Swarrot/Processor/InteropProxyProcessor.php (hunk 1 above)
Comment (makasim) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: The InteropProxyProcessor is useful outside Swarrot, but still using its processors. 
For example you can use it with [enqueue consumption](https://github.com/php-enqueue/enqueue-dev/blob/master/docs/quick_tour.md#consumption) including cli commands or [enqueue simple client](https://github.com/php-enqueue/enqueue-dev/blob/master/docs/quick_tour.md#client)

Comment (olaurendeau) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: Hum, I'm more confident about the Message Provider & Publisher, may the ProxyProcessor can be the object of another PR, so that people can discuss about it ?

Comment (makasim) [inline src/Swarrot/Processor/InteropProxyProcessor.php]: as you wish, gonna remove it

**Comment by makasim** [comment]
@olaurendeau done

**Comment by makasim** [comment]
That would be possible to support routing key once we finalize amqp interop https://github.com/queue-interop/amqp-interop

**Comment by makasim** [comment]
Don't you mind I add Swarrot to the projects list on queue-interop repo? If someone wants to participate I am more than happy to send you an invite. 

Is there any ETA on when this PR is merged?

**Comment by olaurendeau** [comment]
Hello @makasim,
I would have liked another review maybe from @stof or @Taluu 
Let's say that next monday if no other review, I'll merge and release :)

Yes feel free to add swarrot to queue-interop project list :)

**Comment by makasim** [comment]
@olaurendeau thanks for the update. Monday is fine for me.

**Comment by makasim** [comment]
add key arg support, it will work for all amqp interop based transports

**Comment by olaurendeau** [comment]
As soon as travis is happy I'll merge it 👍 
Thanks @makasim !

**Comment by makasim** [comment]
@olaurendeau it's happy

**Comment by olaurendeau** [comment]
It's released @makasim https://github.com/swarrot/swarrot/releases
Thanks a lot for your contribution :)

**Comment by makasim** [comment]
@olaurendeau Would you be interested to join queue interop? I'll be more than happy to send you an invite.

**Comment by olaurendeau** [comment]
Well it would be with pleasure, but I'm not sure to be able to contribute actively, having quite a lot of project running lately :)

**Comment by makasim** [comment]
> but I'm not sure to be able to contribute actively, having quite a lot of project running lately :)

Don't worry.
